Libraries/src/Amazon.Lambda.Annotations/APIGateway/HttpResults.cs

```diff
@@ -0,0 +1,448 @@
+            /// </summary>
+            V2 
+        }
+
+        /// <summary>
+        /// The API Gateway protocol used as the event source. 
+        ///     RestApi -> RestApiAttrbute
+        ///     HttpApi -> HttpApiAttribute
+        /// </summary>
+        public ProtocolFormat Format { get; set; }
+
+        /// <summary>
+        /// The API Gateway protocol version used as the event source. 
+        ///     V1 -> RestApi or HttpApi specifically set as V1
+        ///     V2 -> HttpApi either implicit or explicit set to V2
+        /// </summary>
+        public ProtocolVersion Version { get; set; }
+    }
+
+    /// <summary>
+    /// If this inteface is returned for an API Gateway Lambda function it will serialize itself to the correct JSON format for the 
+    /// configured event source's protocol format and version.
+    /// 
+    /// Users should use the implementation class HttpResults to construct an instance of IHttpResult with the configured, status code, response body and headers.
+    /// </summary>
+    /// <example>
+    /// return HttpResults.Ok("All Good")
+    ///                   .AddHeader("Custom-Header", "FooBar");
+    /// 
+    /// </example>
+    public interface IHttpResult
+    {
+        /// <summary>
+        /// Used by the Lambda Annotations framework to serialize the IHttpResult to the correct JSON response.
+        /// </summary>
+        /// <param name="options"></param>
+        /// <returns></returns>
+        Stream Serialize(HttpResultSerializationOptions options);
+
+        /// <summary>
+        /// Add header to the IHttpResult. The AddHeader method can be called multiple times for the same header to add multi values for a header.
+        /// HTTP header names are case insensitive and the AddHeader method will normalize header name casing by calling ToLower on them.
+        /// </summary>
+        /// <param name="name">HTTP heade rname</param>
+        /// <param name="value">HTTP header value</param>
+        /// <returns>The same instance to allow fluent call pattern.</returns>
+        IHttpResult AddHeader(string name, string value);
+    }
+
+    /// <summary>
+    /// Implementation class for IHttpResult. Consumers should use one of the static methods to create the a result with the desired status code.
+    /// </summary>
+    /// <remarks>
+    /// If a response body is provided it is format using the following rules:
+    /// <list type="bullet">
+    ///     <item>
+    ///         <description>For string then returned as is.</description>
+    ///     </item>
+    ///     <item>
+    ///         <description>For Stream, byte[] or IList&lt;byte&gt; the data is consided binary and base 64 encoded.</description>
+    ///     </item>
+    ///     <item>
+    ///         <description>Anything other type is serialized to JSON.</description>
```

Comment:
I don't have a strong opinion, so this isn't a blocker, but I'm wondering if it'd be helpful if we set `Content-Type:  "application/json"`automatically in this case?

Admittedly when testing in Firefox I was initially wondering _"why doesn't this JSON look normal?"_, which was because it wasn't triggering the JSON viewer.

I haven't tested ASP.NET equivalent to see how it behaves.

Reply:
With some experimenting with ASP.NET Core I added an autofill of content type with the following rules.

* No body - content type not set
* String passed in for body - content type set as `text/plain`
* Stream, Byte[] or IList<byte> - content type set as `application/octet-stream`
* Anything else - content type set to `application/json`

`HttpResults` unit tests updated to reflect this change.

Reply:
Nice, thanks!